hooks/keydata.go

```diff
@@ -0,0 +1,284 @@
+// argument. Whilst this will normally be from [rand.Reader], it can be provided from other
+// secure sources or mocked during tests.
+//
+// The caller can supply a set of snap models and boot modes to bind to the new key.
+//
+// On success, a new key data object is returned, along with primary key and an unlock key
+// that can be added to a storage container.
+func NewProtectedKey(rand io.Reader, params *KeyParams) (protectedKey *secboot.KeyData, primaryKeyOut secboot.PrimaryKey, unlockKey secboot.DiskUnlockKey, err error) {
+	if params == nil {
+		params = new(KeyParams)
+	}
+
+	primaryKey := params.PrimaryKey
+	if len(primaryKey) == 0 {
+		primaryKey = make(secboot.PrimaryKey, 32)
+		if _, err := io.ReadFull(rand, primaryKey); err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot obtain primary key: %w", err)
+		}
+
+	}
+
+	kdfAlg := crypto.SHA256
+	unlockKey, payload, err := secboot.MakeDiskUnlockKey(rand, kdfAlg, primaryKey)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot create new unlock key: %w", err)
+	}
+
+	scope, err := bootscope.NewKeyDataScope(&bootscope.KeyDataScopeParams{
+		PrimaryKey: primaryKey,
+		Role:       params.Role,
+		KDFAlg:     crypto.SHA256,
+		MDAlg:      crypto.SHA256,
+		ModelAlg:   crypto.SHA256,
+	})
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot create boot environment scope: %w", err)
+	}
+
+	if err := scope.SetAuthorizedSnapModels(primaryKey, params.Role, params.AuthorizedSnapModels...); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot set authorized snap models: %w", err)
+	}
+	if err := scope.SetAuthorizedBootModes(primaryKey, params.Role, params.AuthorizedBootModes...); err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot set authorized boot modes: %w", err)
+	}
+
+	aad, err := scope.MakeAEADAdditionalData(secboot.KeyDataGeneration, kdfAlg, secboot.AuthModeNone)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("cannot make AAD: %w", err)
+	}
+
+	var (
+		ciphertext []byte
+		handle     []byte
+		aeadCompat *aeadCompatData
+	)
+
+	keyProtectorMu.Lock()
+	defer keyProtectorMu.Unlock()
+
+	switch {
+	case keyProtectorFlags&KeyProtectorNoAEAD != 0:
+		randBytes := make([]byte, 32+12)
+		if _, err := io.ReadFull(rand, randBytes); err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot obtain random bytes for AEAD compat: %w", err)
+		}
+
+		symKey := randBytes[:32]
+		nonce := randBytes[32:]
+
+		b, err := aes.NewCipher(symKey)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot create cipher for AEAD compat: %w", err)
+		}
+		aead, err := cipher.NewGCM(b)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot create AEAD for AEAD compat: %w", err)
+		}
+		ciphertext = aead.Seal(nil, nonce, payload, aad)
+
+		var encryptedKey []byte
+		encryptedKey, handle, err = keyProtector.ProtectKey(rand, symKey, nil)
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("cannot protect symmetric key for AEAD compat: %w", err)
```

Comment:
it would be good to have a test for the error path coming from the Protector

Reply:
I've added a test for this now.